### ISSUE TITLE
More thinking about default reduction behavior vs slicing 

### ISSUE THREAD

**Issue by DavisVaughan** [DavisVaughan/rray#88]
Julia does the same as I do with reductions, always keep dims https://github.com/JuliaLang/julia/issues/16606 but they argue about it too

**Comment by DavisVaughan** [comment]
The main thing they discuss is that slicing and reductions should have compatible behavior so that they can be immediately broadcast against one another https://groups.google.com/forum/#!topic/julia-users/V-meqUZtb7k

**Comment by DavisVaughan** [comment]
This is a really neat idea from a julia comment.
https://github.com/JuliaLang/julia/issues/16606#issuecomment-398086850

We can't exactly do this because R doesn't allow function overloading like this, but we could either have a separate `squeeze` argument that would error if also set with `axis` or we could make a thin S3 wrapper that would allow `rray_sum(x, axis = squeeze(1, 2))` which would reduce over 1 and 2 but also squeeze the result. Maybe integrate this into array slicing too?

**Comment by DavisVaughan** [comment]
So something like:

``` r
library(rray)
library(magrittr)

x <- matrix(1:6, ncol = 2)

# never drops dimensions, by design
rray_sum(x, axes = 1)
#>      [,1] [,2]
#> [1,]    6   15

# proposed syntax:
# rray_sum(x, axes = squeeze(1))
x %>%
  rray_sum(axes = 1) %>%
  rray_squeeze(axes = 1)
#> [1]  6 15

y <- array(1:8, c(2, 2, 2))

# no dimension dropping
rray_sum(y, axes = 1)
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    3    7
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]   11   15

# drops to 2D
# rray_sum(y, axes = squeeze(1))
y %>%
  rray_sum(1) %>%
  rray_squeeze(1)
#>      [,1] [,2]
#> [1,]    3   11
#> [2,]    7   15
```

But, how would this be different from just having `keep_dims = T/F`? Maybe there would be added benefit if you could do `rray_sum(x, axes = c(1, squeeze(2)))`

**Comment by DavisVaughan** [comment]
Another option is to continue with the default of not dropping dimensions in all cases except `rray_sum(x, axes = NULL)` which reduces over all axes and returns a single number. This would continue to broadcast against anything else correctly, and might be more intuitive. `rray_sum(x, axes = c(1, 2, 3))` would be the way to not drop dimensions but still reduce over all axes (in the case of a 3D array)

**Comment by juangomezduaso** [comment]
I vote for the "keep_dims" attribute :)
 Like the Julia guy said, for simplicity sake
If one is to continue in the multidimensional space I think all dims are to be kept. Even if you want the result in one of the aggregated dimensions, squeezing the others moves all the following axes, including the MARGIN ones. OTOH, for presentation or going to calculate in a lower dimensionality, usually to drop all is what one needs. 
And, in any case, using rray_squeeze() is just one "squeeze axes" reptition away...

**Comment by juangomezduaso** [comment]
But, contrary to base "drop", I think keep_dims=FALSE should preserve all length 1 MARGIN dimensions

**Comment by DavisVaughan** [comment]
> But, contrary to base "drop", I think keep_dims=FALSE should preserve all length 1 MARGIN dimensions

It would definitely do this

**Comment by DavisVaughan** [comment]
How do you feel about this? No `keep_dims` argument, but `axes = NULL` always drops dims because it always results in a length 1 vector, which broadcasts against anything. You can explicitly ask for it to not drop dimensions by setting `axes = c(1, 2)`.

``` r
library(rray)

x <- matrix(1:6, ncol = 2)

rray_sum(x, axes = 1)
#>      [,1] [,2]
#> [1,]    6   15

rray_sum(x, axes = 2)
#>      [,1]
#> [1,]    5
#> [2,]    7
#> [3,]    9

rray_sum(x, axes = c(1, 2))
#>      [,1]
#> [1,]   21

# rray_sum(x, axes = NULL)
# ^ this would return:
rray_squeeze(rray_sum(x))
#> [1] 21
```

I haven't thought it all the way through yet, but it feels like a useful "best of both worlds" compromise. It would mean that `rray_sum()` is not dimensionality stable any more, but so would any other of these other solutions that involve some way to drop dimensions.

This obviously does mean that this is not possible `rray_sum(x, axes = 1, keep_dims = F)`, and instead you have to do `rray_squeeze(rray_sum(x, axes = 1), axes = 1)`, but honestly that might be a good thing.

**Comment by juangomezduaso** [comment]
I totally agree with you. After years of R array programming I am very comfortable with the degenerated (ie length one) dimensions of rray, a perfect fit for broadcasting to keep axes in their positions with no drawbacks. I would only use squeeze mostly for visualization purposes. But in this case, the axes=NULL alternative for scalars and the  same convention in rray_squeeze()'s default axes parameter, make it equaly easy to solve an aditional  lot of situations without any repetition:
``` rray_sum(x, axes = 1) %>% rray_squeeze() ```
So the , after all not so terrible: ``` rray_squeeze(rray_sum(x, axes = 1), axes = 1) ``` would be rarely needed.
